Stop running scripts at shutdown

pnbruckner opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Since the rewrite of the script infrastructure scripts or automation actions, or anything that uses the "script helper", that is running at shutdown can prevent Home Assistant from completing the shutdown process. This change effectively restores the previous behavior that these "action sequences" are stopped at shutdown. It also handles automations triggered at shutdown (or anything else, for that matter, that uses the script helper and runs after shutdown) by giving them up to 60 seconds to complete, and stopping them if they haven't stopped themselves in that period.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

pnbruckner committed:
Stop running scripts at shutdown

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -558,8 +602,23 @@ def __init__(
+
+        all_scripts = hass.data.get("Scripts")
+        if not all_scripts:
+            _HASS = hass
```
No need for a global, we just make sure we pass hass to our async listener

pnbruckner committed:
Update per review

Pass hass to shutdown listeners.

Update script states in first pass shutdown process.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -545,6 +550,41 @@ def _finish(self):
+async def _async_stop_scripts_after_shutdown(hass, point_in_time):
+    """Stop running Script objects started after shutdown."""
+    running_scripts = [
+        script for script in hass.data["Scripts"] if script["instance"].is_running
```
Can we extract this into a constant `DATA_SCRIPTS` and make the value `helpers.script` ?

balloob reviewed (approved):
This looks good. 👍

balloob commented:
When we add tests, we can use `async_fire_time_changed` to advance the time with 60 seconds.

pnbruckner committed:
Update per review 2

balloob commented:
GitHub having issues but verified this all works.
